kani-compiler/src/kani_compiler.rs

```diff
@@ -64,68 +63,262 @@ fn backend(queries: Arc<Mutex<QueryDb>>) -> Box<CodegenBackend> {
-    /// Store the stubs that shall be applied if any.
-    stubs: Option<FxHashMap<DefPathHash, DefPathHash>>,
-    /// Store the arguments for kani compiler.
-    args: Option<ArgMatches>,
+    /// The state which the compiler is at.
+    stage: CompilationStage,
 }
 
 impl KaniCompiler {
     /// Create a new [KaniCompiler] instance.
     pub fn new() -> KaniCompiler {
-        KaniCompiler { queries: QueryDb::new(), stubs: None, args: None }
+        KaniCompiler { queries: QueryDb::new(), stage: CompilationStage::Init }
     }
 
-    /// Method to be invoked after a rustc driver run.
-    /// It will return a list of arguments that should be used in a subsequent call to rustc
-    /// driver. It will return None if it has finished compiling everything.
-    pub fn post_process(&mut self, old_args: Vec<String>) -> Option<Vec<String>> {
-        let stubs = self.stubs.replace(FxHashMap::default()).unwrap_or_default();
-        if stubs.is_empty() {
-            None
+    /// Compile the current crate with the given arguments.
+    ///
+    /// Since harnesses may have different attributes that affect compilation, Kani compiler can
+    /// actually invoke the rust compiler multiple times.
+    pub fn run(&mut self, orig_args: Vec<String>) -> Result<(), ErrorGuaranteed> {
+        while !self.stage.is_done() {
+            debug!(next=?self.stage, "run");
+            match &self.stage {
+                CompilationStage::Init => {
+                    self.run_compilation_session(&orig_args)?;
+                }
+                CompilationStage::CodegenNoStubs { .. } => {
+                    unreachable!("This stage should always run in the same session an Init");
+                }
+                CompilationStage::CodegenWithStubs { target_harness, all_harnesses, .. } => {
+                    let extra_arg =
+                        stubbing::mk_rustc_arg(&all_harnesses[&target_harness].stub_map);
+                    let mut args = orig_args.clone();
+                    args.push(extra_arg);
+                    self.run_compilation_session(&args)?;
+                }
+                CompilationStage::Done => {
+                    unreachable!("There's nothing to be done here.")
+                }
+            };
+
+            self.next_stage();
+        }
+        Ok(())
+    }
+
+    /// Set up the next compilation stage after a `rustc` run.
+    fn next_stage(&mut self) {
+        self.stage = match &mut self.stage {
+            CompilationStage::Init => {
+                // This may occur when user passes arguments like --version or --help.
+                CompilationStage::Done
+            }
+            CompilationStage::CodegenNoStubs { next_harnesses, all_harnesses, .. }
+            | CompilationStage::CodegenWithStubs { next_harnesses, all_harnesses, .. } => {
+                if let Some(target_harness) = next_harnesses.pop() {
+                    CompilationStage::CodegenWithStubs {
```

Comment:
Does it move to the `CodegenWithStubs` stage even if there are no harnesses with stubs?

Reply:
If there is no harness with stubs, the `next_harnesses` vector will be empty, so we will move to Done in the `else` block.

All harnesses without stubs are processed together in one single codegen pass as part of the `CodegenNoStubs` stage.